processlib/activity.py

```diff
@@ -326,58 +349,43 @@ class StartViewActivity(StartMixin, ViewActivity):
 
 
 class EndActivity(Activity):
-    def instantiate(self, **kwargs):
-        super(EndActivity, self).instantiate(**kwargs)
-        self.start()
-        self.finish()
-
-    def finish(self, **kwargs):
-        super(EndActivity, self).finish(**kwargs)
-
-        update_fields = []
-        if not self.process.finished_at:
-            self.process.finished_at = self.instance.finished_at
-            update_fields.append("finished_at")
-
-        if not self.process.status == self.process.STATUS_DONE:
-            self.process.status = self.process.STATUS_DONE
-            update_fields.append("status")
-
-        self.process.save(update_fields=update_fields)
+    def __init__(self, redirect_url_callback=None, completed_message="", **kwargs):
+        super().__init__(**kwargs)
 
+        # Do not use Activity default
+        self.completed_message = completed_message
 
-class EndRedirectActivity(EndActivity):
-    def __init__(self, redirect_url_callback=None, **kwargs):
-        self.redirect_url_callback = redirect_url_callback
-        super(EndActivity, self).__init__(**kwargs)
+        # Retain 'redirect_url_callback' to maintain backward compatibility
+        if redirect_url_callback:
+            self.completed_redirect_url = redirect_url_callback
 
     def instantiate(self, **kwargs):
-        # HACK: we skip the EndActivity implementation
-        # because it would finish the activity right away
-        super(EndActivity, self).instantiate(**kwargs)
-
-    def has_view(self):
-        return True
+        super().instantiate(**kwargs)
 
-    def get_absolute_url(self):
-        return reverse(
-            "processlib:process-activity",
-            kwargs={"flow_label": self.flow.label, "activity_id": self.instance.pk},
-        )
+        if not self.completed_redirect_url:
+            self.start()
+            # Finish activity
+            self.finish(**kwargs)
+            # Finish complete process
+            self.finish_process()
 
     def dispatch(self, request, *args, **kwargs):
         self.start()
-        url = reverse(
-            "processlib:process-detail", kwargs={"pk": self.instance.process.pk}
-        )
-        try:
-            if self.redirect_url_callback:
-                url = self.redirect_url_callback(self)
-            self.finish()
-        except Exception as e:
-            logger.exception(e)
-            self.error(exception=e)
-        return HttpResponseRedirect(url)
+        # Finish activity
```

Comment:
I think having a dispatch here won't work as the activity needs `has_view` true for that to do something.